### PR TITLE
fix torch.linalg.lstsq input check

### LINKED ISSUE #117236: ```torch.linalg.lstsq```: Internal assert failed
### 🐛 Describe the bug

Test code:
```python
import torch

self = torch.randn([1,1,1,1,1], dtype=torch.complex128)
b = torch.randn([1,1,3,1], dtype=torch.complex128)

torch.linalg.lstsq(self, b)
```

Error log
```
Traceback (most recent call last):
  File "/home/yonghyeon/pytorch_err_case_py/lstsq.py", line 6, in <module>
    torch.linalg.lstsq(self, b)
RuntimeError: input.size(-2) == other_2d.size(-2) INTERNAL ASSERT FAILED at "/home/yonghyeon/pytorch/pytorch-asan/aten/src/ATen/native/BatchLinearAlgebra.cpp":3364, please report a bug to PyTorch.
```

Test code (c++):
```cpp
#include <torch/torch.h>

int main() {
    torch::TensorOptions toptions = torch::TensorOptions();

    toptions = toptions.dtype(torch::kComplexDouble);
    torch::Tensor self = torch::randn({1,1,1,1,1}, toptions);

    toptions = toptions.dtype(torch::kComplexDouble);
    torch::Tensor b = torch::randn({1,1,3,1}, toptions);

    torch::linalg::lstsq(self, b, c10::nullopt, c10::nullopt);

    return 0;
}
```

Error log:
```
terminate called after throwing an instance of 'c10::Error'
  what():  input.size(-2) == other_2d.size(-2) INTERNAL ASSERT FAILED at "/home/yonghyeon/pytorch/pytorch-asan/aten/src/ATen/native/BatchLinearAlgebra.cpp":3364, please report a bug to PyTorch. 
Exception raised from linalg_lstsq_out_info at /home/yonghyeon/pytorch/pytorch-asan/aten/src/ATen/native/BatchLinearAlgebra.cpp:3364 (most recent call first):
frame #0: <unknown function> + 0x1e7a6e (0x7f8ac6710a6e in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #1: <unknown function> + 0x1e7a20 (0x7f8ac6710a20 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #2: <unknown function> + 0x1e79b0 (0x7f8ac67109b0 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #3: <unknown function> + 0x1e7860 (0x7f8ac6710860 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #4: <unknown function> + 0x1e9286 (0x7f8ac6712286 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #5: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x1b7 (0x7f8ac670e4d7 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #6: c10::detail::torchCheckFail(char const*, char const*, unsigned int, char const*) + 0x259 (0x7f8ac6707ca9 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libc10.so)
frame #7: ./reproduce_lstsq() [0x40ca5e]
frame #8: <unknown function> + 0x3e5585d (0x7f8aca76485d in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #9: at::native::linalg_lstsq_out(at::Tensor const&, at::Tensor const&, std::optional<double>, std::optional<c10::basic_string_view<char> >, at::Tensor&, at::Tensor&, at::Tensor&, at::Tensor&) + 0x5e5f (0x7f8aca76142f in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #10: <unknown function> + 0x9fb38aa (0x7f8ad08c28aa in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #11: <unknown function> + 0x9fb4832 (0x7f8ad08c3832 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #12: <unknown function> + 0x68902be (0x7f8acd19f2be in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #13: at::_ops::linalg_lstsq_out::call(at::Tensor const&, at::Tensor const&, std::optional<double>, std::optional<c10::basic_string_view<char> >, at::Tensor&, at::Tensor&, at::Tensor&, at::Tensor&) + 0x166e (0x7f8accc7764e in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #14: <unknown function> + 0x3e7b3aa (0x7f8aca78a3aa in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #15: at::native::linalg_lstsq(at::Tensor const&, at::Tensor const&, std::optional<double>, std::optional<c10::basic_string_view<char> >) + 0x1b75 (0x7f8aca76a095 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #16: <unknown function> + 0xa795fef (0x7f8ad10a4fef in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #17: <unknown function> + 0xa796dff (0x7f8ad10a5dff in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #18: <unknown function> + 0x688df20 (0x7f8acd19cf20 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #19: <unknown function> + 0x688d832 (0x7f8acd19c832 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #20: at::_ops::linalg_lstsq::redispatch(c10::DispatchKeySet, at::Tensor const&, at::Tensor const&, std::optional<double>, std::optional<c10::basic_string_view<char> >) + 0x954 (0x7f8acceb1894 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #21: <unknown function> + 0x1192ea2b (0x7f8ad823da2b in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #22: <unknown function> + 0x11658a92 (0x7f8ad7f67a92 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #23: <unknown function> + 0x1165667b (0x7f8ad7f6567b in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #24: <unknown function> + 0x11659875 (0x7f8ad7f68875 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #25: <unknown function> + 0x688df20 (0x7f8acd19cf20 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #26: at::_ops::linalg_lstsq::call(at::Tensor const&, at::Tensor const&, std::optional<double>, std::optional<c10::basic_string_view<char> >) + 0x1319 (0x7f8accc75839 in /home/yonghyeon/pytorch/pytorch-asan/build/lib/libtorch_cpu.so)
frame #27: ./reproduce_lstsq() [0x4162ef]
frame #28: ./reproduce_lstsq() [0x415eaf]
frame #29: ./reproduce_lstsq() [0x40722f]
frame #30: main + 0xaf0 (0x404ff0 in ./reproduce_lstsq)
frame #31: <unknown function> + 0x29d90 (0x7f8abcb65d90 in /lib/x86_64-linux-gnu/libc.so.6)
frame #32: __libc_start_main + 0x80 (0x7f8abcb65e40 in /lib/x86_64-linux-gnu/libc.so.6)
frame #33: _start + 0x25 (0x404435 in ./reproduce_lstsq)

Aborted (core dumped)
```

### Versions

PyTorch version: 2.3.0a0+git03969cb
Is debug build: True
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: 12.0.1
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.12.0 | packaged by Anaconda, Inc. | (main, Oct  2 2023, 17:29:18) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-89-generic-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             96
On-line CPU(s) list:                0-95
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 24
Socket(s):                          2
Stepping:                           7
CPU max MHz:                        4000.0000
CPU min MHz:                        1200.0000
BogoMIPS:                           6000.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          1.5 MiB (48 instances)
L1i cache:                          1.5 MiB (48 instances)
L2 cache:                           48 MiB (48 instances)
L3 cache:                           71.5 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0-23,48-71
NUMA node1 CPU(s):                  24-47,72-95
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.26.2
[pip3] optree==0.10.0
[pip3] torch==2.3.0a0+git03969cb
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-include               2023.1.0         h06a4308_46344  
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] torch                     2.3.0a0+git03969cb           dev_0    <develop>

### PR DESCRIPTION
**Pull request by inkcherry**
Fixes [#117236 ](https://github.com/pytorch/pytorch/issues/117236)
The current case does not meet the vector scenario requirements, and it lacks sufficient checks (relying solely on ```dim_diff``` is insufficient).  Consequently, it triggers an internal assertion error.
cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @lezcano

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
Sounds reasonable. Can you add a regression test?

File: aten/src/ATen/native/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -3524,10 +3524,22 @@ std::tuple<Tensor&, Tensor&, Tensor&, Tensor&> linalg_lstsq_out(
...
+  // auto expected_batched_rhs_shape = IntArrayRef(input.sizes().data(), input.dim() - 1); // input.shape[:-1]
+  // bool vector_case = other.dim() == 1 || (input.dim() - 1 == other.dim() && other.sizes().equals(expected_batched_rhs_shape));
+
+  bool vector_case = linalg_solve_is_vector_rhs(input, other);
+  Tensor other_2d = vector_case ? other.unsqueeze(-1) : other;
   TORCH_CHECK(
       input.size(-2) == other_2d.size(-2),
-      dim_diff ? "torch.linalg.lstsq: input.size(-2) should match other.size(-1)"
+      vector_case ? "torch.linalg.lstsq: input.size(-2) should match other.size(-1)"
                : "torch.linalg.lstsq: input.size(-2) should match other.size(-2)");
```
Comment (nikitaved) [inline aten/src/ATen/native/BatchLinearAlgebra.cpp:3543]: Let's add a test!

Comment (inkcherry) [reply aten/src/ATen/native/BatchLinearAlgebra.cpp:3543]: thanks for the review! added @nikitaved , @lezcano

**Review by lezcano — APPROVED** [review]
Thank you for the fix!

This function could certainly do with a big refactor similar to that of https://github.com/pytorch/pytorch/pull/79300

**Pull request closed without merging**